src/PageHeader/PageHeader.tsx

```diff
@@ -56,19 +57,90 @@ export type PageHeaderProps = {
   as?: React.ElementType | 'header' | 'div'
 } & SxProp
 
-const Root: React.FC<React.PropsWithChildren<PageHeaderProps>> = ({children, sx = {}, as = 'div'}) => {
-  const rootStyles = {
-    display: 'flex',
-    flexDirection: 'column',
-    // TODO: We used hard-coded values for the spacing and font size in this component. Update them to use new design tokens when they are ready to use.
-    gap: '0.5rem',
-  }
-  return (
-    <Box as={as} sx={merge<BetterSystemStyleObject>(rootStyles, sx)}>
-      {children}
-    </Box>
-  )
-}
+const Root = React.forwardRef<HTMLDivElement, React.PropsWithChildren<PageHeaderProps>>(
+  ({children, sx = {}, as = 'div'}, forwardedRef) => {
+    const rootRef = useProvidedRefOrCreate<HTMLDivElement>(forwardedRef as React.RefObject<HTMLDivElement>)
+
+    const [contextArea, setContextArea] = React.useState<string>('')
+    const [titleArea, setTitleArea] = React.useState<string>('leadingAction titleArea trailingAction actions')
+    const [description, setDescription] = React.useState<string>('')
+    const [navigation, setNavigation] = React.useState<string>('')
+
+    const gridTemplateAreas = `${contextArea !== '' ? `'${contextArea}' ` : ''}${titleArea ? `'${titleArea}' ` : ''}${
+      description ? `'${description}' ` : ''
+    }${navigation ? `'${navigation}' ` : ''}`.trim()
+
+    const rootStyles = {
+      display: 'grid',
+      // We have 4 columns.
+      gridTemplateColumns: 'auto auto auto 1fr',
+      // We are rendering the template areas dynamically based on the children's visibility.
+      gridTemplateAreas,
+      gap: 'var(--stack-gap-condensed, 0.5rem)',
+    }
+
+    const [visibility, setVisibility] = React.useState({
+      contextArea: false,
+      titleArea: false,
+      leadingAction: false,
+      trailingAction: false,
+      actions: false,
+      description: false,
+      navigation: false,
+    })
+
+    React.useEffect(() => {
+      const observer = new IntersectionObserver((entries: IntersectionObserverEntry[]) => {
+        for (const entry of entries) {
+          setVisibility(prev => ({
+            ...prev,
+            [entry.target.getAttribute('data-component') as string]: entry.isIntersecting,
+          }))
+        }
+      })
+
+      for (const item of rootRef.current?.children || []) {
+        observer.observe(item)
+      }
+
+      return () => observer.disconnect()
+    }, [rootRef])
+
+    React.useEffect(() => {
+      // reset the grid template area strings when visibility is changed.
```

Comment:
Is there a better/cleaner way to do this? 🤔

Reply:
I'm sure we can figure something out 🙂